docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+	Log logr.Logger
+
+	// Hints is the set of secrets used in the expression(s), which can be used to filter/define what 
+	// if any secrets should be retrieved
+	Hints set.Set[string]
+}
+
+type SecretGetter interface {
+	// GetSecrets retrieves the set of secrets defined by params.
+	// The resulting map is keyed by the secret key (== property name of operatorSpec.secrets.<key>)
+	// which is the same as what is referenced by the user when authoring their expressions and passed in
+	// via params.Hints
+	GetSecrets(
+		ctx context.Context,
+		params *GetSecretsParameters,
+	) (map[string]string, error)
+}
+```
+
+The generic reconciler will check for this extension like it does for others before invoking the new generic code
+to handle the `operatorSpec.formattedSecrets` field.
+
+## Decision
+
+### Aspect 1: Template language
+
+We will use [Option 2: CEL](#option-2-google-common-expression-language-cel) as it gets us the most power for the lowest 
+amount of effort and is fully customizable and already hardened for use in multitenant environments.
+
+### Aspect 2: Location
+
+We will use [Option 2: operatorSpec.formattedConfigMaps and operatorSpec.formattedSecrets](#option-2-two-new-operatorspec-properties-operatorspecformattedsecrets-and-operatorspecformattedconfigmaps).
+It avoids the problems with [Option 3](#option-3-a-new-crd-exporter) and avoids the 
+(admittedly minimal) collision risk with [Option 1](#option-1-part-of-operatorspecsecrets-and-operatorspecconfigmap)
+
+### Aspect 3: Secret access
+
+Only one option was presented and we're using it.
+
+### FAQ
+
+Q: How does this new way avoid calling the secrets APIs if the user didn't ask for any secrets to be exported? 
+Keep in mind that some resources such as ManagedCluster (at v2/api/containerservice/customizations/managed_cluster_extensions.go)
+require the ability to differentiate _which_ secrets are being requested as well because there are multiple
+secrets APIs.
+
+A: For `FormattedSecrets`, the function signature to discover secrets will have `secretHints` passed to it, 
+which will contain details about the specific secrets requested in the formatting expression. The resource-specific 
+implementation can use these hints to determine which (if any) APIs to call. Empty `secretHints` means no secrets
+were requested and no secret APIs should be invoked.
+
+Q: How will `secretHints` be determined, since they're buried in an expression the users authored?
+
+A: The easy (but probably bad) way is to do some basic string-matching for the keywords we're aware of, for example because
+the user will be forced to do something like `secret.foo` in their expression to use the `foo` secret, we can parse
+that out of the expression string manually and build a catalog of what secrets they've requested. The harder but probably
+better way is to examine the CEL AST and find all instances of `SelectExpr` and extract out those that correspond to
+`secret.foo`.
```

Comment:
:+1: for using the CEL AST

Reply:
Yes I looked into how to do it using the CEL AST and it's not very difficult.

Reply:
I'll update this item to just say CEL AST is the answer. It's one of the advantages of using CEL as well.